### DISCUSSION TITLE
Management of Cloud Images and Cloud Specs for K8sCluster

### DISCUSSION THREAD

**Discussion by sykim-etri** [cloud-barista/cb-tumblebug#1665] [Ideas]
컨테이너 인프라인 쿠버네티스 클러스터(K8sCluster)를 위해 활용 가능한 이미지들과 스펙들의 정보 관리 방안에 관해 논의하고자 합니다.

### 확정안

* assets/{cloudimage.csv, cloudspec.csv} 파일의 필드(InfraType) 확장
* 설정값은 vm, instance, k8s, kubernetes, container 등이 가능하며 구분자를 '|' 통해 함께 저장 관리
* 손쉬운 검색을 위해 DB에는 상기 설정값을 유사 키워드로 확장하여 저장/관리
  * vm -> vm 등
  * k8s, kubernetes, container -> k8s|kubernetes|container 등

| cspImageId/cspSpecName | ... | InfraType |
|--------------|---|------------------------------|
| img-01/spec-01 | ... | vm |
| img-01/spec-02 | ... | k8s |
| img-01/spec-03 | ... | vm\|k8s |

### 초안
손쉬운 방안으로, 현재 가상 머신 인프라를 위해 활용 가능한 이미지들과 스펙들의 정보를 관리되는 assets/{cloudimage.csv, cloudspec.csv} 파일의 필드(supportedInfraType(가칭))를 다음과 같이 확장하는 것을 생각해 볼 수 있습니다.

이미지의 경우 가상 머신 인프라와 컨테이너 인프라간 겹치는 경우가 많지 않으나 스펙의 경우 겹치는 경우가 많기 때문에,
동일한 방식 사용을 위해 설정값을 vm(가상 머신용), container(컨테이너용), all(둘다 가능)로 하는 것이 어떨까 합니다.

### assets/cloudimage.csv 확장 방안
| cspImageId | ... | supportedInfraType(가칭) |
|--------------|---|------------------------------|
| img-01 | ... | vm |
| img-02 | ... | container |
| img-03 | ... | all |

### assets/cloudspec.csv 확장 방안
| cspSpecName | ... | supportedInfraType(가칭) |
|-----------------|---|------------------------------|
| spec-01 | ... | vm |
| spec-02 | ... | container |
| spec-03 | ... | all |

검토하시고 좋은 의견 부탁드립니다.
@seokho-son

**Comment by seokho-son** [comment]
@sykim-etri 
제안 감사합니다. 전반적으로 동의 합니다.

supportedInfraType -> infraType

정도로 간결하게 하면 어떨까 싶네요.

- `all` 키워드를 제공하는 경우에는, `all`에 해당하는 것이 무엇인지 확인할 방법이 필요한 것 같긴합니다. conf 를 통해서 지정해두는 것도 방법인 것 같네요.
- 대부분의 spec, image가 vm을 위한 것이라, vm을 일일이 기입하는 것이 불편할 수 있다는 생각이 들긴하는데.. 아무것도 작성하지 않은 경우를 vm으로 처리하는 것은 아무래도 어렵겠지요? 아마도, `container` 용 전용 spec도 있는 것 같으니...

**Reply by sykim-etri** [reply]
Spec 정보를 활용하는 쪽 코드(FilterSpecsByRange)를 다시금 살펴보니 특정 필드의 값을 그대로 조회하는 방식을 취하고 있으므로, csv 파일 업데이트가 조금 번거롭겠지만 vm/container로만 구분하여 중복 항목을 기재하는 방안이 바람직해 보입니다. 모호할 수 있는 all 키워드도 사라지겠습니다.

https://github.com/cloud-barista/cb-tumblebug/blob/70b65fbf530be0f8553602a82ad480f2fe756f0a/src/core/mcir/spec.go#L509-L516

**Reply by seokho-son** [reply]
`all` 등의 처리가 가능은 합니다. asset 에 있는 내용은 그 자체로 검색 테이블이 되지는 않고, 검색용 table을 구성하는 정보값이라고 생각하시면 됩니다.

**Reply by sykim-etri** [reply]
csv 파일 자체가 바로 DB에 반영되진 않긴 하지만 결국 DB에는 vm/container로만 저장해야 할 것으로 보였습니다. 그럼 csv 파일에는 all로 기재하고 DB 구성시 vm/container로 각각 저장하는 방안이 적당할까요?
사용자들이 개별 Spec 등록시에도 csv 파일 포맷을 참조하여 all 키워드를 입력할 수 있을텐데.. 입력 내용과 조회 내용에서의 차이로 혼란이 있을 가능성도 조금 있어 보입니다.

**Reply by sykim-etri** [reply]
RegisterSpecWithCspSpecName()나 RegisterImageWithId()를 통해 등록하는 경우 조금 모호한 상황이 생길 것으로 보입니다. CB-SP로부터 vm/container 정보가 명확하게 구분되어 리턴되지는 않고 KeyValue의 값 중 유추(container 키워드 적용)해 볼 일부 정보만 리턴되는 것으로 보입니다. 해당 키워드가 없다면 기본으로 vm 으로 분류하거나 unknown 등 또 다른 키워드가 필요할 것 같기도 합니다.

**Reply by seokho-son** [reply]
@sykim-etri 넵, 아쉽게도 CB-SP로부터 vm/container 등 여러 정보가 명확하게 구분되어 리턴되지 않는 현황이 맞습니다. 그래서, 현재는 수동으로 입력 중인 상태이고요. CB-SP 응답을 유추해서 임의로 추출해내는 방식은 바람직하지 않다고 생각하여, 가급적 피하고 싶습니다.

우선, assets에는 명시적으로 입력하는 것으로 하고 추진하면 어떨까 싶습니다. (작성이 다소 불편해지긴 하지만)
default에 해당하는 값을 적용할지는 향후 개선사항으로 함께 살펴보시죠. :) 

현재는 csv의 한 아이템 내에서 separator는 통일되어 있지 않은 상태입니다. `,`는 일단 불가하고, 적절히 가독성도 높으면서, 시스템적으로도 오류가 발생할 가능성이 낮은 seperator가 도입되면 좋을 것 같습니다. `/` 도 url 등에서 활용하다보니, 구분자 오류가 발생하는 경우 등이 있는 것 같습니다.
`|` (파이프)를 구분자로 사용하는 것은 어떨까요?

**Reply by seokho-son** [reply]
아, 그리고 vm|container 보다는 vm|k8s 가 적절하지 않을까 싶습니다. (`kubernetes`도 인식하게 해야 함)

**Reply by sykim-etri** [reply]
csv 파일 내에는 '|'를 구분자로 vm, instance, k8s, kubernetes 등으로 입력받고 vm|instance|k8s|kubernetes|container 등으로 확장하여 DB에 저장/관리하는 것으로 정리하겠습니다~

**Reply by seokho-son** [reply]
instance, 는 모호해서~ 일단 제거 부탁드립니닷 ㅎㅎ
